### PR TITLE
[FEATURE] Implement `ConfiguredAssetGCSDataConnector`

### PR DESCRIPTION
**Pull request by cdkini**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- https://superconductive.atlassian.net/secure/RapidBoard.jspa?rapidView=24&projectKey=GREAT&modal=detail&selectedIssue=GREAT-81
- Implement the initial version of the `ConfiguredAssetGCSDataConnector`
- Write a comprehensive suite of unit tests (with mocks) to test that the feature has parity with S3 and Azure

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py
Comment (cdkini) [inline great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py]: Schema validation will ensure that we only have one of these options (can also have neither).

File: great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py
Comment (NathanFarmer) [inline great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py:15]: Nitpick: "Google" instead of "google"

File: great_expectations/datasource/data_connector/util.py
Comment (cdkini) [inline great_expectations/datasource/data_connector/util.py]: This is a little tricky because the value of `delimiter` directly impacts how traversal works in the bucket. If you set it to be `/`, you immediately stop traversal once you finish iterating through the current level.

What do we do if a user asks for a recursive usage but also explicitly passes in a delimiter of `/`? Do we overwrite what the user put in to ensure proper behavior? Or do we leave it as is and have the user get unexpected behavior?

File: great_expectations/datasource/data_connector/util.py
Comment (NathanFarmer) [inline great_expectations/datasource/data_connector/util.py:36]: Same nitpick

File: great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py
Comment (cdkini) [reply great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py:15]: I very much appreciate nitpicks 😄 

Great catch, thank you!

File: great_expectations/datasource/data_connector/util.py
Comment (cdkini) [reply great_expectations/datasource/data_connector/util.py:36]: Fixed both instances!

File: great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py
Comment (Shinnnyshinshin) [inline great_expectations/datasource/data_connector/configured_asset_gcs_data_connector.py:122]: Ah this is how we're doing it :) Love it

File: great_expectations/datasource/data_connector/util.py
Comment (Shinnnyshinshin) [inline great_expectations/datasource/data_connector/util.py:359]: ❤️  thanks for the thorough explanation here

File: great_expectations/datasource/data_connector/util.py
Comment (Shinnnyshinshin) [reply great_expectations/datasource/data_connector/util.py]: Do you think it would help to display a warning to the user? I'm thinking either a logger message or even a print statement

File: tests/datasource/data_connector/test_configured_asset_gcs_data_connector.py
Comment (Shinnnyshinshin) [inline tests/datasource/data_connector/test_configured_asset_gcs_data_connector.py:142]: Thank you for splitting this out in its own fixture. It makes a ton of sense

File: tests/datasource/data_connector/test_configured_asset_gcs_data_connector.py
Comment (Shinnnyshinshin) [inline tests/datasource/data_connector/test_configured_asset_gcs_data_connector.py:754]: ❤️

**Review by Shinnnyshinshin — COMMENTED** [review]
Thank you so much for this @cdkini. I left a comment where you asked a question about `delimiter`, and we can discuss what the best way would be the proceed. Otherwise it looks good to go

File: great_expectations/datasource/data_connector/util.py
Comment (cdkini) [reply great_expectations/datasource/data_connector/util.py]: This is a great idea! Do you think logging or using warning is appropriate here?

**Comment by cdkini** [comment]
Thanks for the review @Shinnnyshinshin! I think you suggestion about logging/warning is a great one.

I've made the following changes from your last review:
1. Added `logging.warning` calls to clarify what is wrong with the user's input and why we're overriding it
2. Beefed up the docstring to better describe the importance of the delimiter and what it means in a Google SDK context.

Let me know if it makes sense or if you think I could further clarify what's going on. Happy to discuss this in further detail tomorrow!

File: great_expectations/datasource/data_connector/util.py
Comment (Shinnnyshinshin) [inline great_expectations/datasource/data_connector/util.py]: 👍🏼

File: great_expectations/datasource/data_connector/util.py
Comment (Shinnnyshinshin) [reply great_expectations/datasource/data_connector/util.py]: That's a good question. between `logger.warning()` and `warnings.warn()` I'm leaning towards `warnings.warn()`, but only because we want to tell the user about potentially unexpected behavior.. but it's your call. I'd support either

**Review by Shinnnyshinshin — APPROVED** [review]
preemptively approving pending your final decision on whether to use `logger.warning()` or `warnings.warn()`. Thank you for this. LGTM!

**Comment by cdkini** [comment]
I think I agree with you here. I'll make the change and then merge.

Thanks for the help as always!

**Pull request merged**
